proxy/ProxyClientSession.cc

```diff
@@ -124,43 +118,35 @@ ProxyClientSession::state_api_callout(int event, void *data)
   case EVENT_NONE:
   case EVENT_INTERVAL:
   case TS_EVENT_HTTP_CONTINUE:
-    if (likely(is_valid_hook(this->api_hookid))) {
-      if (this->api_current == nullptr && this->api_scope == API_HOOK_SCOPE_GLOBAL) {
-        this->api_current = http_global_hooks->get(this->api_hookid);
-        this->api_scope   = API_HOOK_SCOPE_LOCAL;
-      }
+    if (nullptr == cur_hook) {
+      /// Get the next hook to invoke from HttpHookState
+      cur_hook = hook_state.getNext();
+    }
+    if (nullptr != cur_hook) {
+      bool plugin_lock    = false;
+      APIHook const *hook = cur_hook;
 
-      if (this->api_current == nullptr && this->api_scope == API_HOOK_SCOPE_LOCAL) {
-        this->api_current = ssn_hook_get(this->api_hookid);
-        this->api_scope   = API_HOOK_SCOPE_NONE;
-      }
+      Ptr<ProxyMutex> plugin_mutex;
 
-      if (this->api_current) {
-        bool plugin_lock = false;
-        APIHook *hook    = this->api_current;
-        Ptr<ProxyMutex> plugin_mutex;
-
-        if (hook->m_cont->mutex) {
-          plugin_mutex = hook->m_cont->mutex;
-          plugin_lock  = MUTEX_TAKE_TRY_LOCK(hook->m_cont->mutex, mutex->thread_holding);
-          if (!plugin_lock) {
-            SET_HANDLER(&ProxyClientSession::state_api_callout);
-            if (!schedule_event) { // Don't bother to schedule is there is already one out.
-              schedule_event = mutex->thread_holding->schedule_in(this, HRTIME_MSECONDS(10));
-            }
-            return 0;
-          }
-        }
+      if (hook->m_cont->mutex) {
+        plugin_mutex = hook->m_cont->mutex;
+        MUTEX_TRY_LOCK(lock ,hook->m_cont->mutex, mutex->thread_holding);
 
-        this->api_current = this->api_current->next();
-        hook->invoke(eventmap[this->api_hookid], this);
-
-        if (plugin_lock) {
-          Mutex_unlock(plugin_mutex, this_ethread());
+        if (!lock.is_locked()) {
+          SET_HANDLER(&ProxyClientSession::state_api_callout);
+          mutex->thread_holding->schedule_in(this, HRTIME_MSECONDS(10));
+          return 0;
         }
+      }
+
+      cur_hook = nullptr; // mark current callback at dispatched.
+      hook->invoke(eventmap[hook_state.id()], this);
```

Comment:
The presence of eventmap array is very misleading.  This code only ever handles the SSN_START and SSN_CLOSE hooks.  It's OK if you don't want to remove this as a part of this PR.  I'm hoping to get around to removing it eventually.

Reply:
How is it misleading? It translates from the hook ID to the event code.

Reply:
As I said it implies that this code path handles a large number of events rather than just SSN_START and SSN_CLOSE.